include/xsimd/types/xsimd_api.hpp

```diff
@@ -1462,6 +1507,29 @@ namespace xsimd
         return kernel::sadd<A>(B(x), B(y), A {});
     }
 
+    /**
+     * Scatter elements from `src` at addresses starting at
+     * `dst` and offset by each element in `index`.
+     * Scattered elements are stored in order.
+     * (The scale is always 1.)
+     */
+
+    /**
+     * @ingroup batch_data_transfer
+     *
+     * Scatter elements from the batch \c src at addresses starting at \c dst
+     * and offset by each element in \c index. The scale is always 1.
+     * @param src Batch of elements to scatter.
+     * @param dst Destination address
+     * @param index Indexes in which to store the elements to.
+     */
+    template <typename T, typename A, typename U, typename V>
+    inline void scatter(batch<T, A> const& src, U* dst,
```

Comment:
I think You should use ``T == U`` here. The scatter intrinscis you're using don't support another configuration anyway

Reply:
We (Krita) need the source and destination types to be separate, because we need to perform uint8/16<->float32 conversion when gathering/scattering integer depth pixels. These cases are obviously covered by the default, but when the types do match we can let the user take advantage of AVX512F intrinsics.

Reply:
Well, I question the API then. When I read *scatter this register of `uint32_t` to this pointer of `float`*, what kind of conversion happen? Is it a `reinterpret_cast` or a `static_cast`?
If it's a `reinterpret_cast`, then I'd rather have a scatter to `uint32_t*` followed, in client code, by the `reinterpret_cast`.
If it's a `static_cast` well, that's unsupported by the intrinsic and I'd rather have a `batch_cast` followed by a `scatter`.

Just my views on the problem, I'm curious to read your thoughts on this.

Reply:
> Well, I question the API then. When I read scatter this register of uint32_t to this pointer of float, what kind of conversion happen? Is it a reinterpret_cast or a static_cast?

In all cases, where `T` and `U` don't match, it's because we intend xsimd to apply a `static_cast`. (These semantics are taken straight out of Vc's gather, [here](https://vcdevel.github.io/Vc-1.4.1/classVc_1_1Vector.html#aa5788937fe94e616e4587fff8343de53). Same applies for [scatter](https://vcdevel.github.io/Vc-1.4.1/classVc_1_1Vector.html#a877b3f1854117429aca9b4a84561cdbb).)

>  If it's a static_cast well, that's unsupported by the intrinsic and I'd rather have a batch_cast followed by a scatter.

That's not possible (AFAIK). Suppose you need to gather 4 `uint8_t`s into a SSE `batch<float>` (e.g. to calculate a sRGB gaussian mask). `batch_cast` will catch you because of the mismatching vector sizes:

https://github.com/xtensor-stack/xsimd/blob/ce48f25f061dcb2254eb2e75777a792dd73ee9ba/include/xsimd/arch/generic/xsimd_generic_math.hpp#L68

I suppose I could add another instance for where the vector sizes do match?

Reply:
Here's a sampler of four alternatives: template v scratch memory with static_cast v batch_cast. https://godbolt.org/z/zvMdqfzjv

Can't check with Google Benchmark because `DoNotOptimize` seems to not be working correctly at the moment: https://godbolt.org/z/hPbrjfMs9

Reply:
Alternatively, here are sample runs of llvm-mca on the individual alternatives: https://godbolt.org/z/bvjhvcGG4